server/cluster_stat.go

```diff
@@ -0,0 +1,391 @@
+		idx--
+		if idx < 0 {
+			idx += cap
+		}
+	}
+	return usage / float64(steps)
+}
+
+func cpuUsageAll(usages []*pdpb.RecordPair) float64 {
+	sum := 0.0
+	if len(usages) == 0 {
+		return sum
+	}
+	for _, usage := range usages {
+		sum += float64(usage.Value)
+	}
+	return sum / float64(len(usages))
+}
+
+// Keys returns the average written and read keys duration
+// an interval of heartbeats
+func (s *StatEntries) Keys(steps int) (read int64, written int64) {
+	cap := cap(s.entries)
+	if steps > cap {
+		steps = cap
+	}
+	if steps > s.total {
+		steps = s.total
+	}
+	if steps == 0 {
+		return 0, 0
+	}
+
+	idx := (s.total - 1) % cap
+	for i := 0; i < steps; i++ {
+		stat := s.entries[idx]
+		read += int64(stat.KeysRead)
+		written += int64(stat.KeysWritten)
+		idx--
+		if idx < 0 {
+			idx += cap
+		}
+	}
+	return read / int64(steps), written / int64(steps)
+}
+
+// Bytes returns the average written and read bytes duration
+// an interval of heartbeats
+func (s *StatEntries) Bytes(steps int) (read int64, written int64) {
+	cap := cap(s.entries)
+	if steps > cap {
+		steps = cap
+	}
+	if steps > s.total {
+		steps = s.total
+	}
+	if steps == 0 {
+		return 0, 0
+	}
+	idx := (s.total - 1) % cap
+	for i := 0; i < steps; i++ {
+		stat := s.entries[idx]
+		read += int64(stat.BytesRead)
+		written += int64(stat.BytesWritten)
+		idx--
+		if idx < 0 {
+			idx += cap
+		}
+	}
+	return read / int64(steps), written / int64(steps)
+}
+
+// ClusterStatEntries saves the StatEntries for each store in the cluster
+type ClusterStatEntries struct {
+	m        sync.RWMutex
+	stats    map[uint64]*StatEntries
+	size     int   // size of entries to keep for each store
+	interval int64 // average interval of heartbeats
+	total    int64 // total of StatEntry appended
+}
+
+// NewClusterStatEntries returns a statistics object for the cluster
+func NewClusterStatEntries(size int) *ClusterStatEntries {
+	return &ClusterStatEntries{
+		stats: make(map[uint64]*StatEntries),
+		size:  size,
+	}
+}
+
+// Append an store StatEntry
+func (cst *ClusterStatEntries) Append(stat *StatEntry) {
+	cst.m.Lock()
+	defer cst.m.Unlock()
+
+	// update interval
+	start, end := stat.Interval.GetStartTimestamp(),
+		stat.Interval.GetEndTimestamp()
+	interval := int64(end - start)
+	if start == 0 || end == 0 || interval == 0 {
+		interval = DefaultIntervalHint
+	}
+	if cst.total == 0 {
+		cst.interval = interval
+	} else {
+		cst.interval = cst.interval + (interval-cst.interval)/cst.total
+	}
+	cst.total++
+
+	// append the entry
+	storeID := stat.StoreId
+	entries, ok := cst.stats[storeID]
+	if !ok {
+		entries = NewStatEntries(cst.size)
+		cst.stats[storeID] = entries
+	}
+
+	entries.Append(stat)
+}
+
+func contains(slice []uint64, value uint64) bool {
+	for i := range slice {
+		if slice[i] == value {
+			return true
+		}
+	}
+	return false
+}
+
+// CPU returns the cpu usage of the cluster
+func (cst *ClusterStatEntries) CPU(d time.Duration, excludes ...uint64) float64 {
```

Comment:
Add a method to get cpu usage of one store?

Reply:
I don't need it.